<a href="https://colab.research.google.com/github/zodiacfive/InventoryManager/blob/master/SimpleNNTorch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import torch
import torch.nn as nn
import torch.optim as optim

# Define a simple neural network
class SimpleNN(nn.Module):
    def __init__(self, input_size, hidden_size, num_classes):
        super(SimpleNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        self.fc2 = nn.Linear(hidden_size, num_classes)
        # Add Sigmoid activation for binary classification output
      #  self.sigmoid = nn.Sigmoid()

    def forward(self, x):
        out = self.fc1(x)
        out = self.relu(out)
        out = self.fc2(out)
        # Apply Sigmoid to the output
       # out = self.sigmoid(out)
        return out

# Hyperparameters
input_size = 10
hidden_size = 20
num_classes = 2
learning_rate = 0.01
num_epochs = 100

# Dummy data (replace with your actual data)
X = torch.randn(100, input_size)
y = torch.randint(0, num_classes, (100,))

# Instantiate the model
model = SimpleNN(input_size, hidden_size, num_classes)

# Define loss function and optimizer
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

# Training loop
for epoch in range(num_epochs):
    # Forward pass
    outputs = model(X)
    loss = criterion(outputs, y)

    # Backward and optimize
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    if (epoch+1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

print("Training finished!")

Epoch [10/100], Loss: 0.6084
Epoch [20/100], Loss: 0.5267
Epoch [30/100], Loss: 0.4253
Epoch [40/100], Loss: 0.3190
Epoch [50/100], Loss: 0.2246
Epoch [60/100], Loss: 0.1528
Epoch [70/100], Loss: 0.1039
Epoch [80/100], Loss: 0.0718
Epoch [90/100], Loss: 0.0507
Epoch [100/100], Loss: 0.0372
Training finished!


### SimpleNN Model with BCEWithLogitsLoss for Binary Classification

This section demonstrates the SimpleNN model adjusted for binary classification using `BCEWithLogitsLoss`.

In [11]:
import torch
import torch.nn as nn
import torch.optim as optim

# Define a simple neural network (adjusted for binary classification output)
class SimpleNN_Binary(nn.Module):
    def __init__(self, input_size, hidden_size):
        super(SimpleNN_Binary, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        # Output layer for binary classification, 1 output unit
        self.fc2 = nn.Linear(hidden_size, 1)


    def forward(self, x):
        out = self.fc1(x)
        out = self.relu(out)
        # No activation function needed here for BCEWithLogitsLoss
        out = self.fc2(out)
        return out

# Hyperparameters (adjusted num_classes concept for BCEWithLogitsLoss)
input_size = 10
hidden_size = 20
# num_classes is implicitly 1 for the output of the linear layer when using BCEWithLogitsLoss
learning_rate = 0.01
num_epochs = 100

# Dummy data (replace with your actual data)
X = torch.randn(100, input_size)
# BCEWithLogitsLoss expects target labels as float, typically 0.0 or 1.0
y = torch.randint(0, 2, (100,)).float().unsqueeze(1)


# Instantiate the model
# Note: num_classes is handled by the output layer's size (1 in this case)
model_binary = SimpleNN_Binary(input_size, hidden_size)


# Define loss function and optimizer
criterion_binary = nn.BCEWithLogitsLoss()
optimizer_binary = optim.Adam(model_binary.parameters(), lr=learning_rate)

# Training loop
print("Starting training with BCEWithLogitsLoss...")
for epoch in range(num_epochs):
    # Forward pass
    outputs = model_binary(X)
    loss = criterion_binary(outputs, y)

    # Backward and optimize
    optimizer_binary.zero_grad()
    loss.backward()
    optimizer_binary.step()

    if (epoch+1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

print("Training finished with BCEWithLogitsLoss!")

Starting training with BCEWithLogitsLoss...
Epoch [10/100], Loss: 0.6241
Epoch [20/100], Loss: 0.5413
Epoch [30/100], Loss: 0.4554
Epoch [40/100], Loss: 0.3683
Epoch [50/100], Loss: 0.2946
Epoch [60/100], Loss: 0.2352
Epoch [70/100], Loss: 0.1825
Epoch [80/100], Loss: 0.1382
Epoch [90/100], Loss: 0.1052
Epoch [100/100], Loss: 0.0798
Training finished with BCEWithLogitsLoss!


# Task
Implement a neural network for home price prediction using a suitable dataset, including data loading, preprocessing, model definition, training, and evaluation.

## Load and explore data

### Subtask:
Load a suitable home price dataset and perform initial exploration to understand its structure and features.


**Reasoning**:
Load the Boston Housing dataset from scikit-learn, display the first few rows, print column information and data types, and show descriptive statistics.



**Reasoning**:
The Boston Housing dataset is not available and has ethical concerns. Use the California Housing dataset from scikit-learn as an alternative, display the first few rows, print column information and data types, and show descriptive statistics.



In [13]:
from sklearn.datasets import fetch_california_housing
import pandas as pd

# Load the California Housing dataset
housing = fetch_california_housing()

# Create a pandas DataFrame
df = pd.DataFrame(housing.data, columns=housing.feature_names)
df['target'] = housing.target

# Display the first 5 rows
print("First 5 rows of the dataset:")
display(df.head())

# Print column names and their data types
print("\nColumn information and data types:")
df.info()

# Display descriptive statistics of numerical columns
print("\nDescriptive statistics:")
display(df.describe())

First 5 rows of the dataset:


MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  8.3252      41.0  6.984127   1.023810       322.0  2.555556     37.88   
1  8.3014      21.0  6.238137   0.971880      2401.0  2.109842     37.86   
2  7.2574      52.0  8.288136   1.073446       496.0  2.802260     37.85   
3  5.6431      52.0  5.817352   1.073059       558.0  2.547945     37.85   
4  3.8462      52.0  6.281853   1.081081       565.0  2.181467     37.85   

   Longitude  target  
0    -122.23   4.526  
1    -122.22   3.585  
2    -122.24   3.521  
3    -122.25   3.413  
4    -122.25   3.422


Column information and data types:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   MedInc      20640 non-null  float64
 1   HouseAge    20640 non-null  float64
 2   AveRooms    20640 non-null  float64
 3   AveBedrms   20640 non-null  float64
 4   Population  20640 non-null  float64
 5   AveOccup    20640 non-null  float64
 6   Latitude    20640 non-null  float64
 7   Longitude   20640 non-null  float64
 8   target      20640 non-null  float64
dtypes: float64(9)
memory usage: 1.4 MB

Descriptive statistics:


MedInc      HouseAge      AveRooms     AveBedrms    Population  \
count  20640.000000  20640.000000  20640.000000  20640.000000  20640.000000   
mean       3.870671     28.639486      5.429000      1.096675   1425.476744   
std        1.899822     12.585558      2.474173      0.473911   1132.462122   
min        0.499900      1.000000      0.846154      0.333333      3.000000   
25%        2.563400     18.000000      4.440716      1.006079    787.000000   
50%        3.534800     29.000000      5.229129      1.048780   1166.000000   
75%        4.743250     37.000000      6.052381      1.099526   1725.000000   
max       15.000100     52.000000    141.909091     34.066667  35682.000000   

           AveOccup      Latitude     Longitude        target  
count  20640.000000  20640.000000  20640.000000  20640.000000  
mean       3.070655     35.631861   -119.569704      2.068558  
std       10.386050      2.135952      2.003532      1.153956  
min        0.692308     32.540000   -124.350000      0.149990  
25%        2.429741     33.930000   -121.800000      1.196000  
50%        2.818116     34.260000   -118.490000      1.797000  
75%        3.282261     37.710000   -118.010000      2.647250  
max     1243.333333     41.950000   -114.310000      5.000010

## Data preprocessing

### Subtask:
Clean the data, handle missing values, encode categorical features, and scale numerical features as necessary for neural network input.


**Reasoning**:
Check for missing values and data types, separate features and target, scale features, and split the data into training and testing sets.



In [14]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

# 1. Check for missing values
print("Checking for missing values:")
print(df.isnull().sum())

# 2. Identify categorical features
print("\nData types of columns:")
print(df.dtypes)
# Based on the output, there are no categorical features, all are numerical.

# 3. Separate features (X) and target (y)
X = df.drop('target', axis=1)
y = df['target']

# 4. Scale numerical features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)
X_scaled_df = pd.DataFrame(X_scaled, columns=X.columns)
print("\nFirst 5 rows of scaled features:")
display(X_scaled_df.head())

# 5. Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

print("\nShape of training features:", X_train.shape)
print("Shape of testing features:", X_test.shape)
print("Shape of training target:", y_train.shape)
print("Shape of testing target:", y_test.shape)

Checking for missing values:
MedInc        0
HouseAge      0
AveRooms      0
AveBedrms     0
Population    0
AveOccup      0
Latitude      0
Longitude     0
target        0
dtype: int64

Data types of columns:
MedInc        float64
HouseAge      float64
AveRooms      float64
AveBedrms     float64
Population    float64
AveOccup      float64
Latitude      float64
Longitude     float64
target        float64
dtype: object

First 5 rows of scaled features:


MedInc  HouseAge  AveRooms  AveBedrms  Population  AveOccup  Latitude  \
0  2.344766  0.982143  0.628559  -0.153758   -0.974429 -0.049597  1.052548   
1  2.332238 -0.607019  0.327041  -0.263336    0.861439 -0.092512  1.043185   
2  1.782699  1.856182  1.155620  -0.049016   -0.820777 -0.025843  1.038503   
3  0.932968  1.856182  0.156966  -0.049833   -0.766028 -0.050329  1.038503   
4 -0.012881  1.856182  0.344711  -0.032906   -0.759847 -0.085616  1.038503   

   Longitude  
0  -1.327835  
1  -1.322844  
2  -1.332827  
3  -1.337818  
4  -1.337818


Shape of training features: (16512, 8)
Shape of testing features: (4128, 8)
Shape of training target: (16512,)
Shape of testing target: (4128,)


## Define regression model

### Subtask:
Define a neural network architecture appropriate for regression, with an output layer that produces a single continuous value.


**Reasoning**:
Define the neural network architecture for regression as instructed, including the necessary imports and class structure.



In [15]:
import torch.nn as nn

# Define a simple neural network for regression
class RegressionNN(nn.Module):
    def __init__(self, input_size, hidden_size):
        super(RegressionNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)
        self.relu = nn.ReLU()
        # Output layer for regression, 1 output unit for a continuous value
        self.fc2 = nn.Linear(hidden_size, 1)

    def forward(self, x):
        out = self.fc1(x)
        out = self.relu(out)
        out = self.fc2(out)
        return out

print("RegressionNN model defined.")

RegressionNN model defined.


## Compile model

### Subtask:
Compile the model with a suitable optimizer and a loss function for regression (like Mean Squared Error).


**Reasoning**:
Instantiate the regression model and define the loss function and optimizer for training.



In [16]:
import torch.optim as optim
import torch

# Instantiate the RegressionNN model
# X_train has 8 features, so input_size is 8
input_size = X_train.shape[1]
hidden_size = 32 # Chosen hidden layer size
model_regression = RegressionNN(input_size, hidden_size)

# Define the loss function for regression (Mean Squared Error)
criterion_regression = nn.MSELoss()

# Define the optimizer
learning_rate = 0.001
optimizer_regression = optim.Adam(model_regression.parameters(), lr=learning_rate)

print("Regression model instantiated, loss function and optimizer defined.")

Regression model instantiated, loss function and optimizer defined.


## Train model

### Subtask:
Train the neural network on the preprocessed data.


**Reasoning**:
Convert the training data to PyTorch tensors and implement the training loop as described in the instructions.



In [17]:
import torch

# 1. Convert training data to PyTorch tensors
X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
# Unsqueeze y_train to have shape (num_samples, 1) for MSELoss
y_train_tensor = torch.tensor(y_train.values, dtype=torch.float32).unsqueeze(1)

# 2. Define the number of training epochs (already defined in previous steps, but redefining for clarity in this block)
# num_epochs = 100 # Using the previously defined num_epochs

# 3. Implement the training loop
print(f"Starting training for {num_epochs} epochs...")
for epoch in range(num_epochs):
    # Forward pass
    outputs = model_regression(X_train_tensor)
    loss = criterion_regression(outputs, y_train_tensor)

    # Backward and optimize
    optimizer_regression.zero_grad()
    loss.backward()
    optimizer_regression.step()

    # Print loss at regular intervals
    if (epoch + 1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

print("Training finished!")

Starting training for 100 epochs...
Epoch [10/100], Loss: 6.0507
Epoch [20/100], Loss: 5.4960
Epoch [30/100], Loss: 4.9990
Epoch [40/100], Loss: 4.5501
Epoch [50/100], Loss: 4.1399
Epoch [60/100], Loss: 3.7608
Epoch [70/100], Loss: 3.4070
Epoch [80/100], Loss: 3.0749
Epoch [90/100], Loss: 2.7626
Epoch [100/100], Loss: 2.4699
Training finished!


## Evaluate model

### Subtask:
Evaluate the trained model's performance using appropriate regression metrics (e.g., Mean Squared Error, Mean Absolute Error).


**Reasoning**:
Convert test data to tensors, set the model to evaluation mode, calculate predictions, MSE, and MAE, and print the results.



In [18]:
import torch.nn as nn
from sklearn.metrics import mean_absolute_error

# 1. Convert test data to PyTorch tensors
X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test.values, dtype=torch.float32).unsqueeze(1)

# 2. Set the trained model to evaluation mode
model_regression.eval()

# 3. Perform forward pass on test data without gradient calculation
with torch.no_grad():
    predictions = model_regression(X_test_tensor)

# 4. Calculate Mean Squared Error (MSE)
mse = criterion_regression(predictions, y_test_tensor)

# 5. Calculate Mean Absolute Error (MAE)
# Using sklearn's mean_absolute_error for simplicity and clarity
mae = mean_absolute_error(y_test_tensor.numpy(), predictions.numpy())


# 6. Print the calculated MSE and MAE values
print(f"Mean Squared Error (MSE) on test set: {mse.item():.4f}")
print(f"Mean Absolute Error (MAE) on test set: {mae:.4f}")

Mean Squared Error (MSE) on test set: 2.4175
Mean Absolute Error (MAE) on test set: 1.1905


## Make predictions

### Subtask:
Use the trained model to make predictions on new, unseen data.


**Reasoning**:
Create dummy data, set the model to evaluation mode, and make predictions using the trained model.



In [19]:
import torch

# 1. Create a dummy tensor representing new, unseen data
# Ensure it has the same number of features (8) and data type (float32) as training data
num_dummy_samples = 5 # Number of dummy samples
dummy_data = torch.randn(num_dummy_samples, input_size, dtype=torch.float32)

# 2. Set the trained model to evaluation mode
model_regression.eval()

# 3. Use a with torch.no_grad(): block to perform a forward pass and get predictions
with torch.no_grad():
    dummy_predictions = model_regression(dummy_data)

# 4. Print the resulting predictions
print("Predictions on dummy unseen data:")
print(dummy_predictions)

Predictions on dummy unseen data:
tensor([[0.8748],
        [0.9321],
        [0.5859],
        [1.5186],
        [0.4652]])


## Summary:

### Data Analysis Key Findings

*   The California Housing dataset was successfully loaded and explored, revealing 8 numerical features and no missing values.
*   The data was preprocessed by scaling the numerical features using `StandardScaler` and splitting it into training and testing sets (80/20 split).
*   A simple neural network for regression (`RegressionNN`) was defined with an input layer, one hidden layer (size 32), and a single-unit output layer.
*   The model was compiled using the Mean Squared Error (`nn.MSELoss`) as the loss function and the Adam optimizer with a learning rate of 0.001.
*   The neural network was trained for 100 epochs, showing a decrease in training loss from 6.0507 at epoch 10 to 2.4699 at epoch 100.
*   The trained model achieved a Mean Squared Error (MSE) of 2.4175 and a Mean Absolute Error (MAE) of 1.1905 on the test set.
*   The trained model was successfully used to make predictions on dummy unseen data.

### Insights or Next Steps

*   The current model's performance could potentially be improved by experimenting with different network architectures (e.g., adding more layers, changing hidden layer sizes), activation functions, or regularization techniques.
*   Further analysis could involve visualizing the predictions against the actual values on the test set to understand the model's performance distribution and identify potential areas of improvement.
